Added SambaPlayer embed support

lzanol opened the pull request:
This implements SambaPlayer embed support through `<amp-samba-player>` AMP component.

Closes https://github.com/ampproject/amphtml/issues/7755.

lzanol commented:
I signed it!

Cooporation name: SambaTech
Google Groups: https://groups.google.com/forum/#!forum/sambatech

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+/** @private {Object} */
+const API_MAP = {
+  'web4-7091': '//playerapitest2.liquidplatform.com:7091/embed/',
+  'localhost-8080': '//localhost:8080/player-api/embed/',
```
do you need to support localhost and staging for you customers? If this was just for your testing, I recommend only keeping prod configs for the public facing API.

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+
+/** @private {Object} */
+const EVENTS_MAP = {
+  'onPlay': VideoEvents.PLAY,
```
after some manual testing. I don't see `onPlay`, `onMute` or `onUnmute` ever coming from the player. I do see `onStart` and `onResume` and `onPause` however.

aghassemi replied:
Please see test page here: https://amphtml-ae5fc.firebaseapp.com/samba/examples/empty.amp.html 

(keep developer Console open, it prints out events received from the player as you interact with it)

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+    this.env_ = this.element.getAttribute('data-env') || this.env_;
+
+    // the protocol to be used (HTTP, HTTPS)
+    this.protocol_ = (this.element.getAttribute('data-protocol')
```
please only allow HTTPS, I assume all of your endpoints have HTTPS versions. (except for localhost which I think should be removed from the config)

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+      || this.protocol_).replace(/\:?$/, ':');
+
+    // player features related params
+    // WARN: methods missing on returned object (e.g. hasOwnProperty) so recreate it
```
That's by design, why do you need access to `hasOwnProperty`? `getDataParamsFromAttributes` make a pure map so you can do `for in` without needing to check `hasOwnProperty`

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+      params += `${s}${k.replace(/url$/i, 'URL')}=${this.params_[k]}`;
+    }
+
+    params += `&parentURL=#${this.element.ownerDocument.location.href}`;
```
I don't think this is what you want. In most cases this would be `cnd.ampproject.org`, if you want the origin source (e.g. www.myAmpSite.com) use `getSourceOrigin()` method from `url.js`

aghassemi replied:
also `encodeUriComponent` the final value

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+
+    iframe.setAttribute('frameborder', '0');
+    iframe.setAttribute('allowfullscreen', 'true');
+    iframe.src = `${this.protocol_}${API_MAP[this.env_]}${this.projectId_}/`
```
ditto: protocol should not be configurable. always https

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+    const s = '&';
+
+    for (const k in this.params_) {
+      params += `${s}${k.replace(/url$/i, 'URL')}=${this.params_[k]}`;
```
please `encodeUriComponent` both sides of `=`

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+    // not required in case of live
+    this.mediaId_ = this.element.getAttribute('data-media-id') || '';
+
+    // which environment to run
```
please validate "env" too to make sure it is in API_MAP

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+    this.iframe_ = null;
+
+    /** @private {?Function} */
+    this.playerLoadResolver_ = null;
```
no need for the resolver, just need ` this.playerLoadPromise_` (also see my comment on `layoutCallback` regarding this )

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+    return this.loadPromise(iframe)
+      .then(() => {
+        this.element.dispatchCustomEvent(VideoEvents.LOAD);
+        this.playerLoadResolver_(this.iframe_);
```
this can jsut become

```
this.playerLoadPromise_ = this.loadPromise(iframe)
      .then(() => {
         this.element.dispatchCustomEvent(VideoEvents.LOAD);
      });
return  this.playerLoadPromise_;
```

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
```
also set  `this.playerLoadPromise_ = null;`

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+    if (typeof opt_params === 'object') {
+      opt_params = encodeURIComponent(JSON.stringify(opt_params));
+    }
+
```
```
if(!this.playerLoadPromise_ || !this.iframe_) {
  return;
}
```

aghassemi commented on extensions/amp-samba-player/0.1/amp-samba-player.js:
```diff
@@ -0,0 +1,260 @@
+
+  /** @override */
+  play(unusedIsAutoplay) {
+    this.sendMessage_('play');
```
based on my manual testing, none of these actually work. Please see test page at:

https://amphtml-ae5fc.firebaseapp.com/samba/examples/empty.amp.html

aghassemi commented on extensions/amp-samba-player/0.1/validator-amp-samba-player.protoascii:
```diff
@@ -0,0 +1,86 @@
+  attrs: {
+    name: "data-protocol"
+    value_regex_casei: "[0-9a-z]{3,5}"
+  }
```
@honeybadgerdontcare for review

honeybadgerdontcare replied:
@lzanol Thanks for the changes.

@aghassemi validation look good.

aghassemi commented on spec/amp-tag-addendum.md:
```diff
@@ -165,19 +165,20 @@ These may be removed in future versions of AMP
-`<amp-audio>`  
-`<amp-lightbox>`  
+`<amp-ad>`
+`<amp-anim>`
```
don't worry about this file. :) it is super old. we will be removing it.

honeybadgerdontcare commented on extensions/amp-samba-player/0.1/validator-amp-samba-player.protoascii:
```diff
@@ -0,0 +1,86 @@
+# limitations under the license.
+#
+
+tags: {  # amp-samba-player
```
We recently simplified the way to validate extension scripts into a new format. The new format for this tag would look like this:
```
tags: {  # amp-samba-player
  html_format: AMP
  tag_name: "SCRIPT"
  satisfies: "amp-samba-player extension .js script"
  requires: "amp-samba-player"
  extension_spec: {
    name: "amp-samba-player"
    allowed_versions: "0.1"
    allowed_versions: "latest"
  }
  attr_lists: "common-extension-attrs"
}
```

honeybadgerdontcare commented on extensions/amp-samba-player/0.1/validator-amp-samba-player.protoascii:
```diff
@@ -0,0 +1,86 @@
+    value_regex_casei: "[0-9a-z]{3,5}"
+  }
+  attr_lists: "extended-amp-global"
+  spec_url: "https://www.ampproject.org/docs/reference/extended/amp-samba-player.html"
```
The ampproject.org documentation format changed recently. While this would automatically redirect to the new format, lets just use the new format instead:

`spec_url: "https://www.ampproject.org/docs/reference/components/amp-samba-player"`

honeybadgerdontcare commented on extensions/amp-samba-player/0.1/validator-amp-samba-player.protoascii:
```diff
@@ -0,0 +1,86 @@
+  tag_name: "AMP-SAMBA-PLAYER"
+  disallowed_ancestor: "AMP-SIDEBAR"
+  requires: "amp-samba-player extension .js script"
+  attrs: {
```
Could you put these attributes in alphabetical order?

Also these attribute names do not fully match up with what is in the markdown file. E.g. `data-media-protocol` vs `data-protocol` and there is no mention of `data-env` in the markdown.

honeybadgerdontcare commented on extensions/amp-samba-player/0.1/validator-amp-samba-player.protoascii:
```diff
@@ -0,0 +1,86 @@
+  html_format: AMP
+  tag_name: "AMP-SAMBA-PLAYER"
+  disallowed_ancestor: "AMP-SIDEBAR"
+  requires: "amp-samba-player extension .js script"
```
Please add `satisfies: "amp-samba-player"` before the `requires`.

lzanol commented:
Hello @aghassemi and @honeybadgerdontcare, thanks for all the tweaks!

I apologize for the delay, my company has allocated me on another project for a while.

I just did the required adjustments you've pointed out:
- Removed attributes `env` and `protocol` (they were really meant for testing purposes, now always using prod API with HTTPS)
- Removed `this.playerLoadResolver_`
- `encodeURIComponent` on query string keys/values
- Added guard statement on `sendMessage_()`
- `this.playerLoadPromise_ = null` on `unlayoutCallback()`
- Fixed events dispatching/naming
  - Our player has a pending deploy (on coming 2-3 days) for the events `onMute` and `onUnmute` to work. I'll get back to you when it is done.
- Updated and alphabetically ordered custom attributes in ".protoascii" and ".md"

Please let me know if there's something else to adjust.

Best

aghassemi commented:
Hi @lzanol, could you please add amp-samba-player to the video-interface integration tests and make sure they pass?
To do so you need to:
1- Add a test to `test-video-players.js` (similar to amp-youtube that's already there)
2- Include the script in `video-players.html`

These will run an extensive suite of tests to ensure all the postMessages between the component and the iframe work properly. You can run the tests locally by running `gulp test --integration`

If you run into test failures, best way to debug manually is to add autoplay attribute on your component in examples/samba-player.amp.html file and visit localhost:8000/examples/samba-player.amp.max.html (note the .max in the URL, it will ensure local code is used to run the example).

Adding autoplay will instruct AMP runtime to use most of the postMessage video-interface integrations. Your video should start muted auto playing when it comes to view and pause when it goes out of view. Tapping the video should also unmute it.

aghassemi commented:
closing since it is too old, please reopen when planning to resume work.

lzanol commented:
@aghassemi Nice, thanks. I apologize for delaying, we had some other priorities that needed our focus. We're going to priorize AMP as soon as we can, once it's quite meaningful for our clients and we'll reopen this PR.
